scapy/contrib/ikev2.py

```diff
@@ -432,13 +697,46 @@ class IKEv2_payload_CERTREQ(IKEv2_class):
         ]
 
 class IKEv2_payload_CERT(IKEv2_class):
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
+        if _pkt and len(_pkt) >= 16:
+            ts_type = struct.unpack("!B", _pkt[4:5])[0]
+            if ts_type == 4:
+                return IKEv2_payload_CERT_CRT
+            elif ts_type == 7:
+                return IKEv2_payload_CERT_CRL
+            else:
+                return IKEv2_payload_CERT_STR
+        return IKEv2_payload_CERT_STR
+
+class IKEv2_payload_CERT_CRT(IKEv2_payload_CERT):
     name = "IKEv2 Certificate"
     fields_desc = [
         ByteEnumField("next_payload",None,IKEv2_payload_type),
         ByteField("res",0),
-        FieldLenField("length",None,"cert_data","H",adjust=lambda pkt,x:x+5),
+        FieldLenField("length",None,"cert_data","H",adjust=lambda pkt,x: x+len(pkt.x509Cert)+5),
+        ByteEnumField("cert_type",4,IKEv2CertificateEncodings),
+        PacketLenField("x509Cert", X509_Cert(''), X509_Cert, length_from=lambda x:x.length-5),
+        ]
+
+class IKEv2_payload_CERT_CRL(IKEv2_payload_CERT):
+    name = "IKEv2 Certificate"
+    fields_desc = [
+        ByteEnumField("next_payload",None,IKEv2_payload_type),
+        ByteField("res",0),
+        FieldLenField("length",None,"cert_data","H",adjust=lambda pkt,x: x+len(pkt.x509CRL)+5),
+        ByteEnumField("cert_type",7,IKEv2CertificateEncodings),
+        PacketLenField("x509CRL", X509_CRL(''), X509_CRL, length_from=lambda x:x.length-5),
+        ]
+
+class IKEv2_payload_CERT_STR(IKEv2_payload_CERT):
+    name = "IKEv2 Certificate"
+    fields_desc = [
+        ByteEnumField("next_payload",None,IKEv2_payload_type),
+        ByteField("res",0),
+        FieldLenField("length",None,"cert_data","H",adjust=lambda pkt,x: x+5),
         ByteEnumField("cert_type",0,IKEv2CertificateEncodings),
-        StrLenField("cert_data","",length_from=lambda x:x.length-5),
+        StrLenField("cert_data", "", length_from=lambda x:x.length-5),
```

Comment:
If you want to fix missing spaces after commas, you need to do it in a separate commit and also, fix the whole file.